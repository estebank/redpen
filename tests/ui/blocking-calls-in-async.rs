// #![feature(print_internals)]
use std::fs::File;
use std::io::Write;

#[redpen::assume_bad("blocking_async")]
fn interesting() {}
fn uninteresting() {}

#[redpen::assume_ok("blocking_async")]
fn can_block_ignored() {
    println!("can't touch this");
}

fn transitive() {
    interesting();
}

async fn bar() {
    interesting();
}
async fn baz() {
    uninteresting();
}
async fn bal() {
    can_block_ignored();
}
// async fn foo(closure: impl Fn()) {
async fn foo() {
    bar().await;
    // baz().await;
    // println!("Hello, stdout!");
    // interesting();
    // closure();
    // ::std::io::_print(format_args!("Hello, stdout!\n"));
    // let mut file = File::create("foo.txt").unwrap();
    // file.write_all(b"Hello, world!").unwrap();

}
fn bat() {
    interesting();
}

fn main() {
    // let _ = foo(|| { transitive(); });
    let _ = foo();
    
    let _ = bat();
}
