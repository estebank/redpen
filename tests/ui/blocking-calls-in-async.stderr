ERROR rustc_redpen::blocking_async accessee=Spanned { node: UnsizeCast { source_ty: &'{erased} Closure(DefId(1:46 ~ std[c294]::rt::lang_start::{closure#0}), [T/#0, i8, Binder(extern "RustCall" fn(()) -> i32, []), (Binder(fn() -> T/#0, []),)]), target_ty: &'{erased} dyn [Binder(Trait(std::ops::Fn<()>), []), Binder(Projection(ExistentialProjection { def_id: DefId(2:3390 ~ core[719c]::ops::function::FnOnce::Output), args: [()], term: Term::Ty(i32) }), []), Binder(AutoTrait(DefId(2:3139 ~ core[719c]::marker::Sync)), []), Binder(AutoTrait(DefId(2:42098 ~ core[719c]::panic::unwind_safe::RefUnwindSafe)), [])] + '{erased} }, span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:159:9: 159:100 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(1:36 ~ std[c294]::rt::lang_start_internal), [])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:158:17: 163:6 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(1:7369 ~ std[c294]::sys_common::backtrace::__rust_begin_short_backtrace), [Binder(fn() -> T/#0, []), T/#0])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:159:18: 159:82 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(1:5774 ~ std[c294]::process::Termination::report), [T/#0])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:159:18: 159:91 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(0:25 ~ blocking_calls_in_async[202a]::bat), [])), span: $DIR/blocking-calls-in-async.rs:89:5: 89:10 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(0:9 ~ blocking_calls_in_async[202a]::bar), [])), span: $DIR/blocking-calls-in-async.rs:90:13: 90:18 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Drop(Coroutine(DefId(0:10 ~ blocking_calls_in_async[202a]::bar::{closure#0}), [(), std::future::ResumeTy, (), (), CoroutineWitness(DefId(0:10 ~ blocking_calls_in_async[202a]::bar::{closure#0}), []), ()])), span: $DIR/blocking-calls-in-async.rs:90:18: 90:19 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(0:27 ~ blocking_calls_in_async[202a]::interesting), [])), span: $DIR/blocking-calls-in-async.rs:84:5: 84:18 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(2:3391 ~ core[719c]::ops::function::FnOnce::call_once), [F/#0, ()])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:155:18: 155:21 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Drop(T/#1), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:161:1: 161:2 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(2:1737 ~ core[719c]::intrinsics::{extern#2}::black_box), [()])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/hint.rs:338:5: 338:40 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(2:3387 ~ core[719c]::ops::function::FnMut::call_mut), [Self/#0, Args/#1])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Drop(Self/#0), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Drop(Self/#0), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(FnDef(DefId(2:3391 ~ core[719c]::ops::function::FnOnce::call_once), [Self/#0, Args/#1])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71 (#0) }
ERROR rustc_redpen::blocking_async accessee=Spanned { node: Fn(Binder(fn(), [])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71 (#0) }
ERROR rustc_redpen::blocking_async x=(Binder(fn(), []), Fn(Binder(fn(), []))), accessee=Spanned { node: Fn(Binder(fn(), [])), span: /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71 (#0) }
